packages/frontend/src/features/metricsCatalog/hooks/useCatalogCategories.tsx

```diff
@@ -28,11 +37,87 @@ export const useAddCategoryToCatalogItem = () => {
     return useMutation<
         ApiSuccessEmpty['results'],
         ApiError,
-        AddCategoryToCatalogItemParams
+        AddCategoryToCatalogItemParams,
+        {
+            previousCatalog: unknown;
+        }
     >({
         mutationFn: addCategoryToCatalogItem,
-        onSuccess: async () => {
-            await queryClient.invalidateQueries(['metrics-catalog']);
+        onMutate: async ({ catalogSearchUuid, tagUuid, projectUuid }) => {
+            // Cancel any outgoing refetches
+            await queryClient.cancelQueries({
+                queryKey: ['metrics-catalog', projectUuid],
+            });
+
+            // Get the existing tag
+            const existingTag = queryClient
+                .getQueryData<Tag[]>(['project-tags', projectUuid])
+                ?.find((tag) => tag.tagUuid === tagUuid);
+
+            if (!existingTag) return { previousCatalog: null };
+
+            // Get the previous catalog state
+            const previousCatalog = queryClient.getQueryData([
+                'metrics-catalog',
+                projectUuid,
+            ]);
+
+            // Update all matching queries
+            queryClient.setQueriesData<
+                InfiniteData<ApiMetricsCatalog['results']>
+            >(
+                {
+                    queryKey: ['metrics-catalog', projectUuid],
+                    exact: false,
+                },
+                (old) => {
+                    if (!old?.pages) return old;
+
+                    return {
+                        ...old,
+                        pages: old.pages.map((page) => ({
+                            ...page,
+                            data: page.data.map((item) =>
+                                item.catalogSearchUuid === catalogSearchUuid
+                                    ? {
+                                          ...item,
+                                          categories: [
+                                              ...item.categories,
+                                              existingTag,
+                                          ].sort((a, b) =>
+                                              a.name
+                                                  .toLowerCase()
+                                                  .localeCompare(
+                                                      b.name.toLowerCase(),
+                                                  ),
+                                          ),
+                                      }
+                                    : item,
+                            ),
+                        })),
```

Comment:
NP: maybe this can be extracted to a reusable function that receives a mutation callback on the data so that it can used also in update/delete

Reply:
doing it now (separate PR)